From: Wei Yang <richard.weiyang@gmail.com>
Subject: maple_tree: fix mt_destroy_walk() on root leaf node
Date: Wed, 11 Jun 2025 01:12:51 +0000

Patch series "maple_tree: Fix the replacement of a root leaf node", v3.

On destroy we should set each node dead.  But current code miss this when
the maple tree has only the root node.

The reason is that mt_destroy_walk() leverages mte_destroy_descend() to
set the node dead, but this is skipped since the only root node is a leaf.

Patch 1 fixes this.

When adding a test case, I found we always get the new value even when we
leave the old root node not dead.  It turns out we always re-walk the tree
in mas_walk().  It looks like a typo on the status check of mas_walk().

Patch 2 fixes this.

Patch 3 adds a test case to assert retrieving new value when overwriting
the whole range to a tree with only root node.

Without the fix, maple tree user may see old value after overwriting
the whole range [0, ULONG_MAX] on a tree containing just one leaf node.
If it happens, kernel may access some invalid data and crash.

Since we rarely overwrite the whole range of a maple tree, we don't
expect it to have userspace-visible impact in practice.


This patch (of 3):

On destroy, we should set each node dead.  But current code miss this when
the maple tree has only the root node.

The reason is mt_destroy_walk() leverage mte_destroy_descend() to set node
dead, but this is skipped since the only root node is a leaf.

Fixes this by setting the node dead if it is a leaf.

Link: https://lkml.kernel.org/r/20250611011253.19515-1-richard.weiyang@gmail.com
Link: https://lkml.kernel.org/r/20250611011253.19515-2-richard.weiyang@gmail.com
Fixes: 54a611b60590 ("Maple Tree: add new data structure")
Signed-off-by: Wei Yang <richard.weiyang@gmail.com>
Reviewed-by: Liam R. Howlett <Liam.Howlett@Oracle.com>
Cc: Matthew Wilcox (Oracle) <willy@infradead.org>
Cc: <stable@vger.kernel.org>
Signed-off-by: Andrew Morton <akpm@linux-foundation.org>
---

 lib/maple_tree.c |    1 +
 1 file changed, 1 insertion(+)

--- a/lib/maple_tree.c~maple_tree-fix-mt_destroy_walk-on-root-leaf-node
+++ a/lib/maple_tree.c
@@ -5319,6 +5319,7 @@ static void mt_destroy_walk(struct maple
 	struct maple_enode *start;
 
 	if (mte_is_leaf(enode)) {
+		mte_set_node_dead(enode);
 		node->type = mte_node_type(enode);
 		goto free_leaf;
 	}
_
